src/libnnstreamer-edge/nnstreamer-edge-internal.c

```diff
@@ -1368,17 +1391,26 @@ nns_edge_publish (nns_edge_h edge_h, nns_edge_data_h data_h)
   switch (eh->connect_type) {
     case NNS_EDGE_CONNECT_TYPE_TCP:
     case NNS_EDGE_CONNECT_TYPE_HYBRID:
-      conn_data = _nns_edge_get_connection (eh, client_id);
-      if (!conn_data) {
-        nns_edge_loge
-            ("Cannot find connection, invalid client ID or connection closed.");
-        ret = NNS_EDGE_ERROR_INVALID_PARAMETER;
-        break;
+      if(exist_client_id) {
+        *conn_datas = (nns_edge_conn_data_s * ) malloc (sizeof(nns_edge_conn_data_s *) * length);
+        conn_datas[0] = _nns_edge_get_connection (eh, client_id);
+      } else {
+        _nns_edge_get_all_connections(eh, &conn_datas, &length);
       }
 
-      ret = _nns_edge_transfer_data (conn_data->sink_conn, data_h, client_id);
-      if (ret != NNS_EDGE_ERROR_NONE)
-        nns_edge_loge ("Failed to send edge data.");
+      for (i = 0; i < length; i++) {
```

Comment:
not enough time to you :)
recommend to use g_hash_table_get_keys_as_array()
```
int
nns_edge_publish (nns_edge_h edge_h, nns_edge_data_h data_h)
{
  nns_edge_handle_s *eh;
  nns_edge_conn_data_s *conn_data;
  int64_t *client_ids;
  guint i, length;
  char *val;
  int ret = NNS_EDGE_ERROR_NONE;

  eh = (nns_edge_handle_s *) edge_h;
  if (!eh) {
    nns_edge_loge ("Invalid param, given edge handle is null.");
    return NNS_EDGE_ERROR_INVALID_PARAMETER;
  }

  if (nns_edge_data_is_valid (data_h) != NNS_EDGE_ERROR_NONE) {
    nns_edge_loge ("Invalid param, given edge data is invalid.");
    return NNS_EDGE_ERROR_INVALID_PARAMETER;
  }

  nns_edge_lock (eh);

  if (!NNS_EDGE_MAGIC_IS_VALID (eh)) {
    nns_edge_loge ("Invalid param, given edge handle is invalid.");
    nns_edge_unlock (eh);
    return NNS_EDGE_ERROR_INVALID_PARAMETER;
  }

  ret = nns_edge_data_get_info (data_h, "client_id", &val);
  if (ret != NNS_EDGE_ERROR_NONE) {
    nns_edge_logd
        ("Cannot find client ID in edge data. Publish to all connected clients.");
    client_ids = (int64_t *) g_hash_table_get_keys_as_array (eh->conn_table,
        &length);
  } else {
    length = 1U;

    client_ids = (int64_t *) malloc (sizeof (int64_t));
    if (client_ids)
      *client_ids = strtoll (val, NULL, 10);

    SAFE_FREE (val);
  }

  if (!client_ids) {
    nns_edge_loge ("Failed to get client ID, cannot send edge data.");
    goto done;
  }

  /** @todo update code for each connect type */
  switch (eh->connect_type) {
    case NNS_EDGE_CONNECT_TYPE_TCP:
    case NNS_EDGE_CONNECT_TYPE_HYBRID:
      for (i = 0; i < length; i++) {
        conn_data = _nns_edge_get_connection (eh, client_ids[i]);
        if (!conn_data) {
          nns_edge_loge
              ("Cannot find connection, invalid client ID or connection closed.");
          ret = NNS_EDGE_ERROR_INVALID_PARAMETER;
          break;
        }

        ret = _nns_edge_transfer_data (conn_data->sink_conn, data_h,
            client_ids[i]);
        if (ret != NNS_EDGE_ERROR_NONE) {
          nns_edge_loge ("Failed to send edge data.");
          goto done;
        }
      }
      break;
    default:
      break;
  }

done:
  SAFE_FREE (client_ids);
  nns_edge_unlock (eh);
  return ret;
}
```

Reply:
Updated.
Thank you :D